cli_tools/common/utils/logging/firelog_logger.go

```diff
@@ -0,0 +1,311 @@
+	}
+	return strings.Join(anonymizedErrs, "\n")
+}
+
+func getOutputInfo(w *daisy.Workflow, err error) *OutputInfo {
+	o := OutputInfo{
+		TargetSizeGb: getInt64Value(w.GetSerialConsoleOutputValue(targetSizeGb)),
+		SourceSizeGb: getInt64Value(w.GetSerialConsoleOutputValue(sourceSizeGb)),
+	}
+
+	if err != nil {
+		o.FailureMessage = getFailureReason(err)
+		o.FailureMessageWithoutPrivacyInfo = getAnonymizedFailureReason(err)
+	}
+
+	return &o
+}
+
+func getInt64Value(s string) int64 {
+	i, err := strconv.ParseInt(s, 0, 64)
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func (l *FirelogLogger) runWithServerLogging(function func() (*daisy.Workflow, error)) *ComputeImageToolsLogExtension {
+	var event *ComputeImageToolsLogExtension
+
+	// Send log asynchronously. No need to interrupt the main flow when failed to send log, just
+	// keep moving.
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		l.logStart()
+	}()
+
+	if w, err := function(); err != nil {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logFailure(err, w)
+			log.Println(event.OutputInfo.FailureMessage)
+		}()
+	} else {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logSuccess(w)
+		}()
+	}
+
+	wg.Wait()
+	return event
+}
+
+// RunWithServerLogging runs the function with server logging
+func RunWithServerLogging(action string, params InputParams, function func() (*daisy.Workflow, error)) {
+	l := NewFirelogLogger(action, params)
+	l.runWithServerLogging(function)
+}
+
+func (l *FirelogLogger) sendLogToServer(logEvent *ComputeImageToolsLogExtension) logResult {
+	return l.sendLogToServerWithRetry(logEvent, 3)
+}
+
+func (l *FirelogLogger) sendLogToServerWithRetry(logEvent *ComputeImageToolsLogExtension, maxRetry int) logResult {
+	logMutex.Lock()
+	defer logMutex.Unlock()
+
+	for i := 0; i < maxRetry; i++ {
+		// Before sending a new request, wait for a while if server asked to do so
+		if nextRequestWaitMillis > 0 {
+			nextRequestWaitMillis = 0
+			time.Sleep(time.Duration(nextRequestWaitMillis) * time.Millisecond)
+		}
+
+		logRequestJSON, err := constructLogRequest(logEvent)
+		fmt.Println(string(logRequestJSON))
+		if err != nil {
+			fmt.Println("Failed to log to server: failed to prepare json log data.")
+			return failedOnCreateRequestJSON
+		}
+
+		if !serverLogEnabled {
+			return serverLogDisabled
+		}
+
+		req, err := http.NewRequest("POST", l.ServerURL, bytes.NewBuffer(logRequestJSON))
+		if err != nil {
+			fmt.Println("Failed to create log request: ", err)
+			return failedOnCreateRequest
+		}
+
+		req.Header.Set("Content-type", "application/json")
+		req.Header.Set("X-Goog-Api-Key", reverse(key))
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			fmt.Println("Failed to log to server: ", err)
+			continue
+		}
+
+		defer resp.Body.Close()
+		body, _ := ioutil.ReadAll(resp.Body)
+		var lr LogResponse
+		if err = json.Unmarshal(body, &lr); err != nil {
```

Comment:
Will this fail if the server ever returns a response that has more fields than you're prepared for?

Reply:
No, it will be ignored

Reply:
json.Unmarshal will ignore fields that LogResponse doesn't have